### PR TITLE
TimeSeries: Support coloring series and line by thresholds or gradient color scales

### PR DESCRIPTION
**Pull request by torkelo**
* [x] Enables the by value color schemes 
* [x] This enables the "Color series by" option under color scheme (Last | Min | Max) 
* [x] The amazing this just works in legend AND tooltip, hover over green area of line and the tooltip is green hover over red area the tooltip series icon is red, what??? magic :smile:  
* [x] Support for thresholds
* [x] Add support for percentage thresholds
* [x] Support for gradient color schemes 
* [ ] Update docs 
* [x] How to handle negative ranges 
* [x] Test dashboard: http://localhost:3000/d/aBXrJ0R7z/panel-tests-graph-ng-by-value-color-schemes?orgId=1

https://user-images.githubusercontent.com/10999/122473622-6761cd80-cfc2-11eb-85f0-a6d3e1def032.mp4

Screenshot from 2021-06-17 22-55-15

New options
Screenshot from 2021-06-17 22-58-38
Screenshot from 2021-06-17 22-58-53

### CODE REVIEW

**Comment by ryantxu** [comment]
um.... the implementation is shockingly simple

**Comment by torkelo** [comment]
@ryantxu well a lot of the work was done already but disabled, and the things that stumped me before are solved now for some reason. Is the uplot config re-built on every data update? the plot get's new series.color every data update (which it needs for this). 

One thing that this does that is a bit clunky is that if you switch to a Stat/Gauge panel and then back to time series the time series panel will no longer automatically switch to color scheme palettte, but will now be stuck on the "From thresholds" mode. 

We could make it so that if you switch to time series we always clear your color scheme and reset it to Classic palette?

**Comment by dprokop** [comment]
> Is the uplot config re-built on every data update?

@torkelo - no, it is not. It's only re-built when we detect field config/time zone changes.

**Comment by torkelo** [comment]
@dprokop was able to fix it by evaluating color again in legend: https://github.com/grafana/grafana/pull/35910/commits/856b6141456d1ce15afac5ce5677ab96f52b74c2#diff-bae9d583f446dac947d44af6e1fcc179c9f653b55c69cdc194e93d341344f8c2R52

**Comment by dprokop** [comment]
@torkelo I could not get this working with this simple dashboard:

```
{
  "annotations": {
    "list": [
      {
        "builtIn": 1,
        "datasource": "-- Grafana --",
        "enable": true,
        "hide": true,
        "iconColor": "rgba(0, 211, 255, 1)",
        "name": "Annotations & Alerts",
        "type": "dashboard"
      }
    ]
  },
  "editable": true,
  "gnetId": null,
  "graphTooltip": 0,
  "id": 485,
  "links": [],
  "panels": [
    {
      "datasource": null,
      "fieldConfig": {
        "defaults": {
          "color": {
            "mode": "thresholds",
            "seriesBy": "last"
          },
          "custom": {
            "axisLabel": "",
            "axisPlacement": "auto",
            "barAlignment": 0,
            "drawStyle": "line",
            "fillOpacity": 34,
            "gradientMode": "scheme",
            "hideFrom": {
              "legend": false,
              "tooltip": false,
              "viz": false
            },
            "lineInterpolation": "linear",
            "lineWidth": 1,
            "pointSize": 5,
            "scaleDistribution": {
              "type": "linear"
            },
            "showPoints": "auto",
            "spanNulls": false,
            "stacking": {
              "group": "A",
              "mode": "none"
            },
            "thresholdsStyle": {
              "mode": "off"
            }
          },
          "mappings": [],
          "thresholds": {
            "mode": "absolute",
            "steps": [
              {
                "color": "green",
                "value": null
              },
              {
                "color": "red",
                "value": 10
              },
              {
                "color": "#EAB839",
                "value": 20
              },
              {
                "color": "#6ED0E0",
                "value": 50
              },
              {
                "color": "#EF843C",
                "value": 110
              }
            ]
          }
        },
        "overrides": []
      },
      "gridPos": {
        "h": 9,
        "w": 12,
        "x": 0,
        "y": 0
      },
      "id": 2,
      "options": {
        "legend": {
          "calcs": [],
          "displayMode": "list",
          "placement": "bottom"
        },
        "tooltip": {
          "mode": "single"
        }
      },
      "targets": [
        {
          "refId": "A",
          "scenarioId": "csv_metric_values",
          "stringInput": "1,20,90,30,5,0"
        }
      ],
      "title": "Panel Title",
      "type": "timeseries"
    }
  ],
  "refresh": "",
  "schemaVersion": 30,
  "style": "dark",
  "tags": [],
  "templating": {
    "list": []
  },
  "time": {
    "from": "now-6h",
    "to": "now"
  },
  "timepicker": {},
  "timezone": "",
  "title": "Gradient lines",
  "uid": "8MIT9QRnz",
  "version": 4
}
```

Can you guide me what am I doing wrong?

**Comment by torkelo** [comment]
@dprokop pushed two fixes, one fixed the issue you found (with 0 values, caused negative color stop position)

**Comment by dprokop** [comment]
@torkelo indeed, the fix made the dashboard work 🥳 

I have discovered another issue in the attached [1] dashboard. When the series color is changed via the legend (override created), the series disappears completely in the graph.

The confusing thing I see though is the legend icon color itself. While it works in the tooltip perfectly, then in the legend the color of the series is always wrong (it takes the base threshold color I believe). Not sure how we should approach this, maybe we should have a special icon indicating the line color mode is set to thresholds?

[1]
```
{
  "annotations": {
    "list": [
      {
        "builtIn": 1,
        "datasource": "-- Grafana --",
        "enable": true,
        "hide": true,
        "iconColor": "rgba(0, 211, 255, 1)",
        "name": "Annotations & Alerts",
        "type": "dashboard"
      }
    ]
  },
  "editable": true,
  "gnetId": null,
  "graphTooltip": 0,
  "id": 485,
  "links": [],
  "panels": [
    {
      "datasource": null,
      "fieldConfig": {
        "defaults": {
          "color": {
            "mode": "thresholds",
            "seriesBy": "last"
          },
          "custom": {
            "axisLabel": "",
            "axisPlacement": "auto",
            "barAlignment": 0,
            "drawStyle": "line",
            "fillOpacity": 0,
            "gradientMode": "scheme",
            "hideFrom": {
              "legend": false,
              "tooltip": false,
              "viz": false
            },
            "lineInterpolation": "smooth",
            "lineWidth": 1,
            "pointSize": 5,
            "scaleDistribution": {
              "type": "linear"
            },
            "showPoints": "auto",
            "spanNulls": false,
            "stacking": {
              "group": "A",
              "mode": "none"
            },
            "thresholdsStyle": {
              "mode": "off"
            }
          },
          "mappings": [],
          "thresholds": {
            "mode": "absolute",
            "steps": [
              {
                "color": "green",
                "value": null
              },
              {
                "color": "red",
                "value": 10
              },
              {
                "color": "#EAB839",
                "value": 20
              },
              {
                "color": "#6ED0E0",
                "value": 50
              },
              {
                "color": "#EF843C",
                "value": 110
              }
            ]
          }
        },
        "overrides": [
          {
            "matcher": {
              "id": "byName",
              "options": "B-series"
            },
            "properties": [
              {
                "id": "color",
                "value": {
                  "fixedColor": "purple",
                  "mode": "fixed"
                }
              }
            ]
          }
        ]
      },
      "gridPos": {
        "h": 9,
        "w": 12,
        "x": 0,
        "y": 0
      },
      "id": 2,
      "options": {
        "legend": {
          "calcs": [],
          "displayMode": "list",
          "placement": "bottom"
        },
        "tooltip": {
          "mode": "single"
        }
      },
      "targets": [
        {
          "refId": "A",
          "scenarioId": "csv_metric_values",
          "stringInput": "1,20,90,30,5,0"
        },
        {
          "hide": false,
          "max": 100,
          "min": 2,
          "noise": 0,
          "refId": "B",
          "scenarioId": "random_walk",
          "spread": 10
        }
      ],
      "title": "Panel Title",
      "type": "timeseries"
    }
  ],
  "refresh": "",
  "schemaVersion": 30,
  "style": "dark",
  "tags": [],
  "templating": {
    "list": []
  },
  "time": {
    "from": "now-6h",
    "to": "now"
  },
  "timepicker": {},
  "timezone": "",
  "title": "Gradient lines",
  "uid": "8MIT9QRnz",
  "version": 6
}
```

File: packages/grafana-ui/src/components/uPlot/config/gradientFills.ts [hunk 1]
```diff
@@ -57,35 +58,53 @@ export function getScaleGradientFn(
   }
 
   return (plot: uPlot, seriesIdx: number) => {
+    // A uplot bug (I think) where this is called before there is bbox
```
Comment (leeoniya) [inline packages/grafana-ui/src/components/uPlot/config/gradientFills.ts:61]: @torkelo will check this out today

Comment (torkelo) [reply packages/grafana-ui/src/components/uPlot/config/gradientFills.ts:61]: @leeoniya it only happens for the color used by hover focus point

File: packages/grafana-ui/src/components/uPlot/config/gradientFills.ts [hunk 2]
```diff
@@ -57,35 +58,67 @@ export function getScaleGradientFn(
...
+      // when above range we get negative values here
+      if (pos < 0) {
+        return;
+      }
 
-    const getColorWithAlpha = (color: string) => {
-      return 'rgb(255, 0, 0)';
-    };
+      const percent = Math.max(pos / canvasHeight, 0);
+      const realColor = tinycolor(theme.visualization.getColorByName(color)).setAlpha(opacity).toString();
```
Comment (leeoniya) [inline packages/grafana-ui/src/components/uPlot/config/gradientFills.ts:85]: this runs on every data update, and for typical cases (5s refresh) should be ok, but makes me nervous for streaming. would be nice to have a uniform #rgb or #rgba format so we can use the much cheaper colorManipulator#alpha()

https://github.com/grafana/grafana/pull/35911#discussion_r654189622

File: packages/grafana-ui/src/components/uPlot/config/gradientFills.ts (hunk 1 above)
Comment (leeoniya) [reply packages/grafana-ui/src/components/uPlot/config/gradientFills.ts:61]: i'll experiment with calling this lazily for hover points, and maybe let's try with using a CSS color gradient for this. (`cursor.points` can already have their own `.fill` and `.stroke` override callbacks which are expected to return css-compatible `background` and `border` values)

**Review by leeoniya — APPROVED** [review]

**Pull request merged**